### PR TITLE
[JENKINS-38048] Permit credentialsId dropdowns to be used from a global configuration screen

### PR DESCRIPTION
**Pull request by jglick**
[JENKINS-38048](https://issues.jenkins-ci.org/browse/JENKINS-38048)

@reviewbybees esp. @stephenc

### CODE REVIEW

**Comment by jglick** [comment]
Complements #433.

**Comment by stephenc** [comment]
:bee:

**Comment by jglick** [comment]
@reviewbybees done

**Comment by MarkEWaite** [comment]
Thanks. I'm deep in items at work today, then plan to spend much of the weekend preparing for the release of git plugin 3.0 and git client plugin 2.0 (add submodule authentication, switch to JDK 7, depend on Jenkins 1.625 or newer, etc.).  Can this wait until after Jenkins World, or do you need this (and #433) prior to Jenkins World next week?

I may be able to evaluate this during the hackfest late next week, if others don't mind me focusing on the git plugin and git client plugin during the hackfest.

**Comment by jglick** [comment]
> do you need this (and #433) prior to Jenkins World

Well, if you happened to release #433 very soon, I might be able to incorporate it into demos of external Pipeline libraries, but I was certainly not relying on it. There is a fallback implementation (_Legacy SCM_) which works with existing Git plugin releases, just not quite as nicely.

See you soon!

File: src/main/java/hudson/plugins/git/UserRemoteConfig.java [hunk 1]
```diff
@@ -185,7 +188,7 @@ public FormValidation doCheckUrl(@AncestorInPath Item item,
             return FormValidation.ok();
         }
 
-        private static StandardCredentials lookupCredentials(Item project, String credentialId, String uri) {
+        private static StandardCredentials lookupCredentials(@CheckForNull Item project, String credentialId, String uri) {
             return (credentialId == null) ? null : CredentialsMatchers.firstOrNull(
                         CredentialsProvider.lookupCredentials(StandardCredentials.class, project, ACL.SYSTEM,
```
Comment (stephenc) [inline src/main/java/hudson/plugins/git/UserRemoteConfig.java]: I'd rather be explicit in plugins about what to fall back to if the project is `null`... perhaps we should discuss this and https://github.com/jenkinsci/git-plugin/pull/437

Comment (jglick) [inline src/main/java/hudson/plugins/git/UserRemoteConfig.java]: `lookupCredentials` already accepted null. Perhaps you meant to comment on a different line?

**Comment by jglick** [comment]
@MarkEWaite claims this patch is not working for him, so I will merge against current `master` and try to add test coverage.

**Comment by jglick** [comment]
BTW the branching strategy in this plugin has left me a bit confused, since  `master` is creating `2.6.1-SNAPSHOT`; but there is a 3.0.0 (nonbeta) release and a `3.0.0-beta` branch. Assuming for now that this is all irrelevant for purposes of this PR.

**Comment by MarkEWaite** [comment]
You're very polite to call the current branching of the plugin confusing.  My apologies.  3.0.0-beta branch is (for the moment) the most recently released branch on the stream.  

Once PR433 and PR437 have been merged to the master and to 3.0.0-beta, I intend to change the branching so that the master branch will match with the 3.0.0-beta branch and will again be the branch used for future releases of the plugin

The 2.6.x branch will be the branch from which any releases of the 2.6 version of the plugin are released.

Sorry for the confusion created by the my not making the branch transition on the git plugin that I've already made on the git client plugin.

**Comment by jglick** [comment]
No problem. Sounds like it is fine for this to continue to be based on `master`.

I cannot reproduce any problem after merging with `master`, updating `credentials` (to 2.1.7), and running with `workflow-cps-global-lib`. The credentials I have defined:

creds

are available as expected:

config

Nonetheless I will do a bit more interactive testing and add a functional test.

File: src/main/java/hudson/plugins/git/UserRemoteConfig.java [hunk 2]
```diff
@@ -85,7 +85,8 @@ public String toString() {
         public ListBoxModel doFillCredentialsIdItems(@AncestorInPath Item project,
                                                      @QueryParameter String url,
                                                      @QueryParameter String credentialsId) {
-            if (project == null || !project.hasPermission(Item.EXTENDED_READ)) {
+            if (project == null && !Jenkins.getActiveInstance().hasPermission(Jenkins.ADMINISTER) ||
+                project != null && !project.hasPermission(Item.EXTENDED_READ)) {
```
Comment (jglick) [inline src/main/java/hudson/plugins/git/UserRemoteConfig.java]: This is the diff hunk currently covered by the functional test—the one most important for the reported bug.

**Comment by jglick** [comment]
Looks like I need to resolve some merge conflict here; will try to do it soon.

**Comment by MarkEWaite** [comment]
I think I've already resolved them in the version I'm testing.  It is a simple change to UserRemoteConfig.java

**Comment by MarkEWaite** [comment]
Merged.

File: src/main/java/hudson/plugins/git/UserRemoteConfig.java [hunk 3]
```diff
@@ -82,10 +83,12 @@ public String toString() {
     @Extension
     public static class DescriptorImpl extends Descriptor<UserRemoteConfig> {
 
+        @SuppressFBWarnings(value="NP_NULL_PARAM_DEREF", justification="pending https://github.com/jenkinsci/credentials-plugin/pull/68")
```
Comment (jglick) [inline src/main/java/hudson/plugins/git/UserRemoteConfig.java]: Note that this was merged so this comment could be amended to say that the suppression could be deleted as of `credentials` 2.1.9.

File: src/main/java/jenkins/plugins/git/GitSCMSource.java [hunk 4]
```diff
@@ -189,10 +190,12 @@ public String getDisplayName() {
             return Messages.GitSCMSource_DisplayName();
         }
 
+        @SuppressFBWarnings(value="NP_NULL_PARAM_DEREF", justification="pending https://github.com/jenkinsci/credentials-plugin/pull/68")
```
Comment (jglick) [inline src/main/java/jenkins/plugins/git/GitSCMSource.java]: Ditto here of course.

**Pull request merged**